warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java

```diff
@@ -32,174 +33,303 @@
+   * @param proximityInZoneMaxMeanSpeed Minimum mean speed in the proximityZone to prevent a split, in m/s
+   * @param splitTypeLabel              If set, add a label with the reason of the split
+   * @param stoppedTimeLabel            If set, add a label with the time in the stop state
    * @return
    * @throws WarpScriptException
    */
-  private static List<GeoTimeSerie> motionSplit(GeoTimeSerie gts, String label, long timeThreshold, double distanceThreshold, double proximityZoneRadius, double proximityZoneSpeed, long proximityZoneTime) throws WarpScriptException {
+  private static List<GeoTimeSerie> motionSplit(GeoTimeSerie gts, String startLabel, long timeThreshold, double distanceThreshold, double proximityZoneRadius, double proximityZoneMaxSpeed,
+                                                long proximityZoneTime, double proximityInZoneMaxMeanSpeed, String splitTypeLabel, String stoppedTimeLabel) throws WarpScriptException {
     //
     // Sort GTS according to timestamps
     //
-    
+
     GTSHelper.sort(gts, false);
 
     GeoTimeSerie split = null;
-    
+
     List<GeoTimeSerie> splits = new ArrayList<GeoTimeSerie>();
-    
+
     int idx = 0;
-    
+
     int n = GTSHelper.nvalues(gts);
-    
+
     boolean mustSplit = true;
 
     long refLocation = GeoTimeSerie.NO_LOCATION;
     long refTimestamp = Long.MIN_VALUE;
-    
+
     long previousTimestamp = Long.MIN_VALUE;
     long previousLocation = GeoTimeSerie.NO_LOCATION;
     double proximityZoneTraveledDistance = 0.0D;
-    
+
     while (idx < n) {
       long timestamp = GTSHelper.tickAtIndex(gts, idx);
       long location = GTSHelper.locationAtIndex(gts, idx);
       long elevation = GTSHelper.elevationAtIndex(gts, idx);
       Object value = GTSHelper.valueAtIndex(gts, idx);
-    
+
+      //
+      // Initialise refLocation as soon as there is a valid position
+      //
+      if (GeoTimeSerie.NO_LOCATION == refLocation && GeoTimeSerie.NO_LOCATION != location) {
+        refLocation = location;
+        refTimestamp = timestamp;
+      }
+
+      //
+      // If the previous tick was more than 'timeThreshold' ago, split now
+      //
+      if (timestamp - previousTimestamp > timeThreshold) {
+        if (null != splitTypeLabel) {
+          split.getMetadata().putToLabels(splitTypeLabel, SPLIT_TYPE_TIME);
+        }
+        mustSplit = true;
+      }
+
       //
-      // If the previous tick was more than 'timeThreashold' ago, split now
       // If the distance to the previous location is above distanceThreshold, split
       //
-      
-      if (timestamp - previousTimestamp > timeThreshold
-          || (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold)) {
+      if (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold) {
```

Comment:
If the previous check lead to a split, should this one override the splitting reason?

Reply:
Yes. In the current code, priority is:
Stopped>distance split>time split>end of sequence

Do you want to change this ?

Reply:
I want to keep the priority of "stopped" detection over time split or distance split. That is why I do not do a if(time split) else if (distance split) else if (stop detection)

We can implement a DISTANCESPLIT function to have a fast one.